src/compiler/program.ts

```diff
@@ -430,13 +430,14 @@ namespace ts {
         return program;
 
         function getCommonSourceDirectory() {
-            if (typeof commonSourceDirectory === "undefined") {
-                if (options.rootDir && checkSourceFilesBelongToPath(files, options.rootDir)) {
+            if (commonSourceDirectory === undefined) {
+                const emittedFiles = filterSourceFilesInDirectory(files, isSourceFileFromExternalLibrary);

```

Comment:
OK, I added a test `commonSourceDirectory_dts.ts `. This appears to already be the behavior of the current compiler -- the referenced file does not become part of the output.